### PR TITLE
[DPE-2140] Juju 3.x secrets -- running on Juju 3.x

### PR DESCRIPTION
**Pull request by juditnovak**
This branch (PR) only exists to execute the pipeline for Juju 3.1.2.0 (that is not yet used in the "official" MongoDB charm pipeline)

([The PR to be merged](https://github.com/canonical/mongodb-k8s-operator/pull/161) is equal to the current one, except the last commit.)

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -88,7 +90,7 @@ jobs:
           fi
       - name: Run integration tests
         # set a predictable model name so it can be consumed by charm-logdump-action
-        run: sg microk8s -c "tox run -e ${{ matrix.tox-environments }} -- --model testing -m '${{ steps.select-tests.outputs.mark_expression }}'"
+        run: sg snap_microk8s -c "tox run -e ${{ matrix.tox-environments }} -- --model testing -m '${{ steps.select-tests.outputs.mark_expression }}'"
         env:
           CI_PACKED_CHARMS: ${{ needs.build.outputs.charms }}
       - name: Dump logs
```
Comment (juditnovak) [inline .github/workflows/ci.yaml:96]: These changes are NOT to be merged. This PR is only to run the code on 3.1 pipelines

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    (i.e. not the real juju version)
+    """
+    if version("juju") < "3":
+        mocker.patch.object(
+            JujuVersion, "has_secrets", new_callable=PropertyMock
+        ).return_value = False
+    else:
+        mocker.patch.object(
+            JujuVersion, "has_secrets", new_callable=PropertyMock
+        ).return_value = True
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply tests/conftest.py:25]: (May be particularly useful if we plan to have 'matrix' pipelines both for Juju 2.x and 3.x)

File: tox.ini [hunk 3]
```diff
@@ -128,7 +128,7 @@ pass_env =
     CI_PACKED_CHARMS
 commands_pre =
     poetry install --with integration
-    poetry run pip install juju==2.9.42.1
+    poetry run pip install juju==3.1.2.0
 commands =
     poetry run pytest -v --tb native --log-cli-level=INFO -s --durations=0 {posargs} {[vars]tests_path}/integration/test_teardown.py
```
Comment (juditnovak) [inline tox.ini:133]: Again, the `tox.ini` changs are NOT part of the PR -- were only needed for the pipeline. Will NOT be merged.

([The PR that IS to be merged](https://github.com/canonical/mongodb-k8s-operator/pull/161) is equal to the current one, except the last commit.)

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (juditnovak) [reply .github/workflows/ci.yaml:96]: ([The PR to be merged](https://github.com/canonical/mongodb-k8s-operator/pull/161) is equal to the current one, except the last commit.)

File: src/charm.py [hunk 4]
```diff
@@ -614,29 +650,87 @@ def _set_leader_unit_active_if_needed(self):
...
+            )
         else:
-            raise RuntimeError("Unknown secret scope.")
+            scope_obj = self._scope_opj(scope)
+            secret = scope_obj.add_secret({key: value}, label=f"{scope}:{key}")
+            if not secret:
+                raise SecretNotAddedError(f"Couldn't set secret {key}")
+            logging.info(
+                f"Secret {key}, published to Juju Secret Store with label "
+                f"{scope}:{key}, ID: {secret.id} label: {secret.label}"
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply src/charm.py]: Mistake on my side, removed already

File: src/charm.py [hunk 5]
```diff
@@ -650,21 +744,23 @@ def restart_mongod_service(self):
...
-            Config.CONF_DIR + "/" + Config.TLS.KEY_FILE_NAME,
-            keyfile,  # type: ignore
-            make_dirs=True,
-            permissions=0o400,
-            user=Config.UNIX_USER,
-            group=Config.UNIX_GROUP,
-        )
+        keyfile = self.get_secret(APP_SCOPE, SECRET_KEYFILE_NAME)
+        if not keyfile:
+            logger.info(f"No secret defined for {APP_SCOPE}, keyfile")
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply src/charm.py]: We were thinking of the same thing in the same time :-)
IT's been turned into an exception since.

File: src/charm.py [hunk 6]
```diff
@@ -712,10 +851,16 @@ def delete_tls_certificate_from_workload(self) -> None:
...
+        for file in [
+            Config.TLS.EXT_CA_FILE,
+            Config.TLS.EXT_PEM_FILE,
+            Config.TLS.INT_CA_FILE,
+            Config.TLS.INT_PEM_FILE,
+        ]:
+            try:
+                container.remove_path(f"{Config.CONF_DIR}/{file}")
+            except (PathError, ProtocolError, MissingSecretError) as e:
+                logger.error("Cannot put keyFile: %r", e)
```
Comment (marceloneppel) [inline src/charm.py:863]: Is it retried at some point?

Comment (marceloneppel) [inline src/charm.py:863]:
```suggestion
                container.remove_path(f"{Config.CONF_DIR}/{file}")
            except (PathError, ProtocolError, MissingSecretError) as e:
                logger.error("Cannot remove file: %r", e)
```

File: src/charm.py [hunk 7]
```diff
@@ -614,29 +661,118 @@ def _set_leader_unit_active_if_needed(self):
...
-            self.unit_peer_data.update({key: value})
-        elif scope == "app":
-            if not value:
-                del self.app_peer_data[key]
-                return
-            self.app_peer_data.update({key: value})
+        return secret.id
+
+    def set_secret(self, scope: Scopes, key: str, value: Optional[str]) -> Optional[str]:
+        """Get TLS secret from the secret storage."""
```
Comment (marceloneppel) [inline src/charm.py:743]:
```suggestion
        """Set secret into the secret storage."""
```

File: src/charm.py [hunk 8]
```diff
@@ -614,29 +661,118 @@ def _set_leader_unit_active_if_needed(self):
         ):
             self.unit.status = ActiveStatus()
 
-    def get_secret(self, scope: str, key: str) -> Optional[str]:
+    @staticmethod
+    def _mongo_secret_label(scope: Scopes, key: str) -> str:
+        return f"{scope}:{key}"
```
Comment (marceloneppel) [inline src/charm.py:666]: Nice idea!

File: src/charm.py [hunk 9]
```diff
@@ -614,29 +650,87 @@ def _set_leader_unit_active_if_needed(self):
...
-            raise RuntimeError("Unknown secret scope.")
+            scope_obj = self._scope_opj(scope)
+            secret = scope_obj.add_secret({key: value}, label=f"{scope}:{key}")
+            if not secret:
+                raise SecretNotAddedError(f"Couldn't set secret {key}")
+            logging.info(
+                f"Secret {key}, published to Juju Secret Store with label "
+                f"{scope}:{key}, ID: {secret.id} label: {secret.label}"
+            )
+            peer_data.update({key: secret.label})
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply src/charm.py]: You're right, let's do it for sure. May matter for Observer tracking

File: src/exceptions.py [hunk 10]
```diff
@@ -4,13 +4,37 @@
...
     """Raised when a commands to create an admin user on MongoDB fail."""
 
     pass
 
 
-class ApplicationHostNotFoundError(Exception):
+class ApplicationHostNotFoundError(MongoError):
     """Raised when a queried host is not in the application peers or the current host."""
 
     pass
```
Comment (juditnovak) [inline src/exceptions.py:22]: NOTE: The changes above do not relate to the PR (and I can roll them back in case it may be an issue).
Since I've implemented secrets-related exceptions accordingly, I altered the existing errors for consistency.
Note that the change is seamless and fully backwards compatible.

***Rationale:***
I find it very useful when group of exceptions can be caught alltogether, when we don't want to handle a specific issue. Typically for usages such as

```
try:
    connect_to_whatever_db()
except WhateverDBError:
    # We don't care for the specifics this time, but only that the error occured on a service level
    logging.error("DB connection failed")
except NetworkError:
    logging.error("A network issue occured")   
```
rather than
```
try:
    connect_to_whatever_db()
except (
        WhateverDBErrorConnectionError,
        WhateverDBErrorAuthenticationError,
        WhateverDBErrorInvalidRoleErrorm,
#        IForgotYetAnotherWhateverDBConnectionRelatedErrrorToCatchSoItBubblesUp
):
    logging.error("Connection failed")
```
(Typically, in our case: is it a MongoDB error? Or is it an error on the level of our charm?)

File: tests/integration/test_charm.py [hunk 11]
```diff
@@ -148,6 +237,7 @@ async def test_scale_up(ops_test: OpsTest):
     )
 
 
+@pytest.mark.usefixtures("only_without_juju_secrets")
```
Comment (juditnovak) [inline tests/integration/test_charm.py:240]: As indicated in the corresponding, specific commit: this is a temporary change until https://github.com/canonical/mongodb-k8s-operator/pull/176 is merged. I'll rebase on that, wanna have relevant pipelines until.

File: tests/integration/ha_tests/test_ha.py [hunk 12]
```diff
@@ -140,6 +140,7 @@ async def test_scale_up_capablities(ops_test: OpsTest, continuous_writes) -> Non
     await verify_writes(ops_test)
 
 
+@pytest.mark.usefixtures("only_without_juju_secrets")
```
Comment (juditnovak) [inline tests/integration/ha_tests/test_ha.py:143]: As indicated in the corresponding, specific commit: this is a temporary change until https://github.com/canonical/mongodb-k8s-operator/pull/176 is merged. I'll rebase on that, wanna have relevant pipelines until.

File: src/charm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (juditnovak) [reply src/charm.py]: We do that now, on secret creation.

File: src/charm.py [hunk 13]
```diff
@@ -744,12 +935,15 @@ def push_tls_certificate_to_workload(self) -> None:
 
     def delete_tls_certificate_from_workload(self) -> None:
         """Deletes certificate from the workload container."""
-        logger.error("Deleting TLS certificate from workload container")
+        logger.info("Deleting TLS certificate from workload container")
```
Comment (juditnovak) [inline src/charm.py:938]: Changing this back to error (just don't wanna interfere with ongoing pipelines.)

**Review by delgod — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 13 above)
Comment (delgod) [reply src/charm.py:926]: I think "info" is more relevant

File: src/charm.py [hunk 14]
```diff
@@ -631,29 +695,158 @@ def _set_leader_unit_active_if_needed(self):
...
+            return secret.get_content()
+        except ModelError:
+            tmp_secret = Secret(
+                id=self.secrets[scope][SECRET_LABEL].id, backend=self.model._backend
+            )
+            # No label is submtted, as due to a bug it triggers
+            # https://github.com/juju/juju/blob/e826f8b6c9be599ccc63aec79b4993f543cc68e6/worker/uniter/runner/context/context.go#L861
+            # as get_content() is "forcing" the usage both of labels and IDs
+            # https://github.com/canonical/operator/blob/6fcda007bff2671fa97924fbf70f480753f862d6/ops/model.py#L1096
+            return tmp_secret.get_content()
```
Comment (delgod) [inline src/charm.py:724]: Lets always fetch by id.

File: src/charm.py (hunk 13 above)
Comment (juditnovak) [reply src/charm.py:903]: Not changing it then :-)

File: src/charm.py [hunk 15]
```diff
@@ -631,29 +695,158 @@ def _set_leader_unit_active_if_needed(self):
...
+        if not secret:
+            logging.error(f"Secret {scope}:{key} wasn't deleted: no secrets are available")
+            return
+
+        secret_cache = self.secrets[scope].get(SECRET_CACHE_LABEL)
+        if not secret_cache or key not in secret_cache:
+            logging.error(f"No secret {scope}:{key}")
+            return
+
+        secret_cache[key] = SECRET_DELETED_LABEL
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply src/charm.py]: Yes, it is: https://warthogs.atlassian.net/browse/DPE-2267

**Review by delgod — APPROVED** [review]

File: src/charm.py [hunk 16]
```diff
@@ -631,29 +694,138 @@ def _set_leader_unit_active_if_needed(self):
...
+        # We need to create a brand-new secret for this scope
+        else:
+            scope_obj = self._scope_opj(scope)
+
+            secret = scope_obj.add_secret({key: value})
+            if not secret:
+                raise SecretNotAddedError(f"Couldn't set secret {scope}:{key}")
+
+            self.secrets[scope][Config.Secrets.SECRET_LABEL] = secret
+            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply src/charm.py:782]: Hmmm, wait... 
Shouldn't we just rather do:
```
self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL] = {key: value}
```
here to avoid fetching? (Or did I get something wrong? `get_content()` should return us `{key: value}` at this point --assuming no stale read ;-)  Why to fetch at all, instead of using what's already available here?)

File: tests/integration/test_charm.py [hunk 17]
```diff
@@ -113,6 +116,90 @@ async def test_monitor_user(ops_test: OpsTest) -> None:
...
+
+    password = "blablabla"
+    await set_password(ops_test, unit_id=non_leaders[0], username="monitor", password=password)
+    password1 = await get_password(ops_test, unit_id=leader_id, username="monitor")
+    assert password1 != password
+
+    await set_password(ops_test, unit_id=leader_id, username="monitor", password=password)
+    for unit_id in UNIT_IDS:
+        password = await get_password(ops_test, unit_id=unit_id, username="monitor")
+        assert password == password
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply tests/integration/test_charm.py:134]: :facepalm:

File: src/charm.py [hunk 18]
```diff
@@ -399,12 +435,35 @@ def _on_set_password(self, event: ActionEvent) -> None:
...
+
+    def _on_secret_remove(self, event):
+        logging.debug(f"Secret {event._id} seems to have no observers, could be removed")
+
+    def _on_secret_changed(self, event):
+        secret = event.secret
+
+        if secret.id in self.app_peer_data:
+            scope = APP_SCOPE
+        elif secret.id in self.unit_peer_data:
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply src/charm.py:457]: I'm really concerned why this didn't break any tests. This is a proper functional error :-(

File: lib/charms/mongodb/v0/mongodb_tls.py [hunk 19]
```diff
@@ -177,41 +190,49 @@ def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
             event.defer()
             return
 
-        logger.debug("Restarting mongod with TLS enabled.")
+        logger.info("Restarting mongod with TLS enabled.")
 
+        # We may wanna add this?
+        # self.charm.delete_tls_certificate_from_workload()
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/mongodb/v0/mongodb_tls.py]: OK. I add this as a separate PR https://warthogs.atlassian.net/browse/DPE-2281

File: src/charm.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (delgod) [reply src/charm.py]: yes, LGTM

**Review by delgod — DISMISSED** [review]

**Comment by juditnovak** [comment]
The "real" PR https://github.com/canonical/mongodb-k8s-operator/pull/161 is merged, this one was only here for discussions.
